### ISSUE TITLE
[Bug] train_model() sklearn pipeline is trained and cached at every cold start but never used for prediction — wastes 20-30s startup time with dead code

### ISSUE THREAD

**Issue by anshul23102** [Arnav-Singh-5080/IPL-Win-Probability-Prediction#99]
## Description

`application.py` calls `train_model()` at module load time, which reads both CSV files, engineers all features, and fits a full XGBoost pipeline inside a `@st.cache_resource` block:

```python
# application.py line ~1076
@st.cache_resource
def train_model():
    matches = pd.read_csv("matches.csv")
    deliveries = pd.read_csv("deliveries.csv")
    # ... full feature engineering ...
    pipe = Pipeline([
        ('preprocessor', preprocessor),
        ('model', XGBClassifier(n_estimators=100, ...))
    ])
    pipe.fit(X, y)
    return pipe

pipe = train_model()  # Called at startup — takes ~20-30s on cold start
```

However, `pipe` is **never referenced again** in the entire file. When the user clicks **Run Analysis**, the prediction code bypasses `pipe` entirely and loads a completely different model from disk:

```python
# application.py — inside the `if analyze:` block
loaded_xgb = xgb.XGBClassifier()
loaded_xgb.load_model('xgb_win_prob_model.json')  # Loads separate file every click
```

The two models are also incompatible — `train_model()` uses sklearn feature names (`target`, `wickets`) while the prediction manually builds a feature dict with different names (`target_score`, `wickets_left`). Passing `pipe.predict()` the prediction input would raise a feature name mismatch error.

## Impact

1. **Startup performance**: Every Streamlit cold start (server restart, Streamlit Cloud deployment, sleep-wake cycle) triggers a full model train on ~750k delivery rows. This adds 20-30 seconds of blocking startup time that provides zero benefit to the user.
2. **Dead code**: `pipe` is assigned but never read. Any future contributor who attempts to use `pipe` for predictions will get incorrect results due to the feature name mismatch.
3. **Double resource usage**: The app loads both the sklearn pipeline into memory AND the XGBoost JSON model, consuming roughly double the RAM needed.

## Steps to Reproduce

1. Restart the Streamlit server (or clear the cache with `st.cache_resource.clear()`).
2. Open the app.
3. Observe the ~20-30 second startup delay.
4. Search the entire `application.py` file for usages of the variable `pipe` — it is assigned once and never read.

## Proposed Fix

Remove `train_model()` entirely (or keep it only in the notebook for experimentation). At startup, load the pre-trained `xgb_win_prob_model.json` once using `@st.cache_resource`:

```python
@st.cache_resource
def load_model():
    model = xgb.XGBClassifier()
    model.load_model('xgb_win_prob_model.json')
    return model

loaded_xgb = load_model()  # Cached — loaded once, reused on every prediction
```

Then in the prediction block, replace:
```python
# Remove these two lines:
loaded_xgb = xgb.XGBClassifier()
loaded_xgb.load_model('xgb_win_prob_model.json')
```
with the cached `loaded_xgb` from above.

This reduces cold start time from ~30s to under 1 second and eliminates the dead training code.

## Affected File

`application.py` — `train_model()` function (lines ~1076-1125) and the prediction block (lines ~1522-1524)

## Suggested Labels

`type:bug` `level:advanced` `quality:exceptional`

I would like to work on this issue under GSSoC 2026. Could you please assign it to me? @Arnav-Singh-5080

/assign

**Comment by anshul23102** [comment]
Opened PR #109 with the fix.

Root cause: `@st.cache_resource` on `train_model()` meant every cold start executed the full training pipeline (CSV reads + feature engineering + model fit) before any user interaction, even though `pipe.pkl` was already committed and `joblib` was in requirements.

Three changes:
- Removed `@st.cache_resource` from `train_model()` and marked it as an offline utility for regenerating `pipe.pkl` only
- Added `@st.cache_resource load_pipeline()` as the sole runtime entry point: loads `pipe.pkl` via `joblib` in under a second; if the artifact is absent, falls back to `train_model()` once and auto-saves `pipe.pkl` so all future cold starts use the fast path automatically
- Replaced `pipe = train_model()` with `pipe = load_pipeline()`
